### PR TITLE
#165 Support Dagger's assisted inject and generate necessary code

### PR DESCRIPTION
**Pull request by vRallev**
Dagger 2.31 has built-in support for assisted injection. Dagger generates code for the classes using the new annotations. Anvil allows to replace the Dagger annotation processor in many modules and generates the necessary code. This PR adds support for assisted injection.

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (FrostRocket) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:65]: This method does a lot. I feel like it would be beneficial to try to split it up for unit test and readability purposes.

**Review by FrostRocket — APPROVED** [review]
There's a lot here to review, I'd get another pair of eyes on this one before you merge. Overall, flow looks good aside from the factory generator method being a bit verbose. To be fair, there's a lot of logic that it needs to do, though. It's up to you if you wish to take another stab at refactoring.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:65]: I tried to do that in the past and it didn't work well, because there are many variables that need to be shared. Most of it is also a giant builder. Unit tests also verify that the whole output matches the expectation for a given input. It's not worth to test single steps.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (christiandeange) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:202]: Change to
```kotlin
addStatement("return %L.get($argumentList)", delegateFactoryName)
```

or simply:
```kotlin
addStatement("return ${delegateFactoryName}.get($argumentList)")
```

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (christiandeange) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:224]: super nit: you have a `.let { addType(it) }` below that is all on a single line, yet this one is multiline. Make them consistent.

_(Same in `AssistedInjectGenerator`)_

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt
Comment (christiandeange) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt:30]: There's no comments in this file...may want to add some to document what's going on.

File: integration-tests/library/src/main/java/com/squareup/anvil/test/AssistedService.kt
Comment (christiandeange) [inline integration-tests/library/src/main/java/com/squareup/anvil/test/AssistedService.kt:9]: nit: could you use a separate, custom class as the injected type? Something like even a `data class Name(val fullName: String)` might make the example a bit more clear, as injecting a raw `String` isn't a super-typical use case.

**Review by christiandeange — APPROVED** [review]
I have no experience with AssistedInject so I really am not familiar with the implementation much...to be frank I glossed over a large part of the implementation.

Happy to give a ✅ but if you want some more confidence, I would ask for another reviewer that is familiar with the tool you're trying to replace here.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:92]: This error message is coming from Dagger. I kept the same wording so that Anvil can be a one to one replacement.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:128]: I like that, but will keep it as is to avoid allocating extra objects.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:202]: The 2nd, good catch.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt:30]: Added KDoc.

File: integration-tests/library/src/main/java/com/squareup/anvil/test/AssistedService.kt
Comment (vRallev) [reply integration-tests/library/src/main/java/com/squareup/anvil/test/AssistedService.kt:9]: This really just a tiny integration test and doesn't serve as a sample. An extra class is unnecessary and doesn't make it clearer IMO.

**Pull request merged**
